File: programs/mango-v4/src/instructions/perp_create_market.rs
use anchor_lang::prelude::*;
use fixed::types::I80F48;

use crate::accounts_zerocopy::AccountInfoRef;
use crate::error::*;
use crate::state::*;
use crate::util::fill_from_str;

use crate::accounts_ix::*;
use crate::logs::{emit_stack, PerpMarketMetaDataLog};

#[allow(clippy::too_many_arguments)]
pub fn perp_create_market(
    ctx: Context<PerpCreateMarket>,
    perp_market_index: PerpMarketIndex,
    settle_token_index: TokenIndex,
    name: String,
    oracle_config: OracleConfigParams,
    base_decimals: u8,
    quote_lot_size: i64,
    base_lot_size: i64,
    maint_base_asset_weight: f32,
    init_base_asset_weight: f32,
    maint_base_liab_weight: f32,
    init_base_liab_weight: f32,
    maint_overall_asset_weight: f32,
    init_overall_asset_weight: f32,
    base_liquidation_fee: f32,
    maker_fee: f32,
    taker_fee: f32,
    min_funding: f32,
    max_funding: f32,
    impact_quantity: i64,
    group_insurance_fund: bool,
    fee_penalty: f32,
    settle_fee_flat: f32,
    settle_fee_amount_threshold: f32,
    settle_fee_fraction_low_health: f32,
    settle_pnl_limit_factor: f32,
    settle_pnl_limit_window_size_ts: u64,
    positive_pnl_liquidation_fee: f32,
    platform_liquidation_fee: f32,
) -> Result<()> {
    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();

    let mut perp_market = ctx.accounts.perp_market.load_init()?;
    *perp_market = PerpMarket {
        group: ctx.accounts.group.key(),
        settle_token_index,
        perp_market_index,
        blocked1: 0,
        group_insurance_fund: u8::from(group_insurance_fund),
        bump: *ctx.bumps.get("perp_market").ok_or(MangoError::SomeError)?,
        base_decimals,
        name: fill_from_str(&name)?,
        bids: ctx.accounts.bids.key(),
        asks: ctx.accounts.asks.key(),
        event_queue: ctx.accounts.event_queue.key(),
        oracle: ctx.accounts.oracle.key(),
        oracle_config: oracle_config.to_oracle_config(),
        stable_price_model: StablePriceModel::default(),
        quote_lot_size,
        base_lot_size,
        maint_base_asset_weight: I80F48::from_num(maint_base_asset_weight),
        init_base_asset_weight: I80F48::from_num(init_base_asset_weight),
        maint_base_liab_weight: I80F48::from_num(maint_base_liab_weight),
        init_base_liab_weight: I80F48::from_num(init_base_liab_weight),
        open_interest: 0,
        seq_num: 0,
        registration_time: now_ts,
        min_funding: I80F48::from_num(min_funding),
        max_funding: I80F48::from_num(max_funding),
        impact_quantity,
        long_funding: I80F48::ZERO,
        short_funding: I80F48::ZERO,
        funding_last_updated: now_ts,
        base_liquidation_fee: I80F48::from_num(base_liquidation_fee),
        maker_fee: I80F48::from_num(maker_fee),
        taker_fee: I80F48::from_num(taker_fee),
        fees_accrued: I80F48::ZERO,
        fees_settled: I80F48::ZERO,
        fee_penalty,
        settle_fee_flat,
        settle_fee_amount_threshold,
        settle_fee_fraction_low_health,
        settle_pnl_limit_factor,
        padding3: Default::default(),
        settle_pnl_limit_window_size_ts,
        reduce_only: 0,
        force_close: 0,
        padding4: Default::default(),
        maint_overall_asset_weight: I80F48::from_num(maint_overall_asset_weight),
        init_overall_asset_weight: I80F48::from_num(init_overall_asset_weight),
        positive_pnl_liquidation_fee: I80F48::from_num(positive_pnl_liquidation_fee),
        fees_withdrawn: 0,
        platform_liquidation_fee: I80F48::from_num(platform_liquidation_fee),
        accrued_liquidation_fees: I80F48::ZERO,
        reserved: [0; 1848],
    };

    let oracle_ref = &AccountInfoRef::borrow(ctx.accounts.oracle.as_ref())?;
    if let Ok(oracle_price) =
        perp_market.oracle_price(&OracleAccountInfos::from_reader(oracle_ref), None)
    {
        perp_market
            .stable_price_model
            .reset_to_price(oracle_price.to_num(), now_ts);
    } else {
        perp_market.stable_price_model.reset_on_nonzero_price = 1;
    }

    let mut orderbook = Orderbook {
        bids: ctx.accounts.bids.load_init()?,
        asks: ctx.accounts.asks.load_init()?,
    };
    orderbook.init();

    emit_stack(PerpMarketMetaDataLog {
        mango_group: ctx.accounts.group.key(),
        perp_market: ctx.accounts.perp_market.key(),
        perp_market_index,
        base_decimals,
        base_lot_size,
        quote_lot_size,
        oracle: ctx.accounts.oracle.key(),
    });

    Ok(())
}

File: py/src/instructions/perp_create_market.py
from __future__ import annotations
import typing
from solana.publickey import PublicKey
from solana.transaction import TransactionInstruction, AccountMeta
import borsh_construct as borsh
from .. import types
from ..program_id import PROGRAM_ID


class PerpCreateMarketArgs(typing.TypedDict):
    perp_market_index: int
    name: str
    oracle_config: types.oracle_config.OracleConfig
    base_decimals: int
    quote_lot_size: int
    base_lot_size: int
    maint_asset_weight: float
    init_asset_weight: float
    maint_liab_weight: float
    init_liab_weight: float
    liquidation_fee: float
    maker_fee: float
    taker_fee: float
    min_funding: float
    max_funding: float
    impact_quantity: int
    group_insurance_fund: bool
    trusted_market: bool
    fee_penalty: float
    settle_fee_flat: float
    settle_fee_amount_threshold: float
    settle_fee_fraction_low_health: float
    settle_token_index: int


layout = borsh.CStruct(
    "perp_market_index" / borsh.U16,
    "name" / borsh.String,
    "oracle_config" / types.oracle_config.OracleConfig.layout,
    "base_decimals" / borsh.U8,
    "quote_lot_size" / borsh.I64,
    "base_lot_size" / borsh.I64,
    "maint_asset_weight" / borsh.F32,
    "init_asset_weight" / borsh.F32,
    "maint_liab_weight" / borsh.F32,
    "init_liab_weight" / borsh.F32,
    "liquidation_fee" / borsh.F32,
    "maker_fee" / borsh.F32,
    "taker_fee" / borsh.F32,
    "min_funding" / borsh.F32,
    "max_funding" / borsh.F32,
    "impact_quantity" / borsh.I64,
    "group_insurance_fund" / borsh.Bool,
    "trusted_market" / borsh.Bool,
    "fee_penalty" / borsh.F32,
    "settle_fee_flat" / borsh.F32,
    "settle_fee_amount_threshold" / borsh.F32,
    "settle_fee_fraction_low_health" / borsh.F32,
    "settle_token_index" / borsh.U16,
)


class PerpCreateMarketAccounts(typing.TypedDict):
    group: PublicKey
    admin: PublicKey
    oracle: PublicKey
    perp_market: PublicKey
    bids: PublicKey
    asks: PublicKey
    event_queue: PublicKey
    payer: PublicKey
    system_program: PublicKey


def perp_create_market(
    args: PerpCreateMarketArgs,
    accounts: PerpCreateMarketAccounts,
    program_id: PublicKey = PROGRAM_ID,
    remaining_accounts: typing.Optional[typing.List[AccountMeta]] = None,
) -> TransactionInstruction:
    keys: list[AccountMeta] = [
        AccountMeta(pubkey=accounts["group"], is_signer=False, is_writable=False),
        AccountMeta(pubkey=accounts["admin"], is_signer=True, is_writable=False),
        AccountMeta(pubkey=accounts["oracle"], is_signer=False, is_writable=False),
        AccountMeta(pubkey=accounts["perp_market"], is_signer=False, is_writable=True),
        AccountMeta(pubkey=accounts["bids"], is_signer=False, is_writable=True),
        AccountMeta(pubkey=accounts["asks"], is_signer=False, is_writable=True),
        AccountMeta(pubkey=accounts["event_queue"], is_signer=False, is_writable=True),
        AccountMeta(pubkey=accounts["payer"], is_signer=True, is_writable=True),
        AccountMeta(
            pubkey=accounts["system_program"], is_signer=False, is_writable=False
        ),
    ]
    if remaining_accounts is not None:
        keys += remaining_accounts
    identifier = b']/-\xc3>\xfc "'
    encoded_args = layout.build(
        {
            "perp_market_index": args["perp_market_index"],
            "name": args["name"],
            "oracle_config": args["oracle_config"].to_encodable(),
            "base_decimals": args["base_decimals"],
            "quote_lot_size": args["quote_lot_size"],
            "base_lot_size": args["base_lot_size"],
            "maint_asset_weight": args["maint_asset_weight"],
            "init_asset_weight": args["init_asset_weight"],
            "maint_liab_weight": args["maint_liab_weight"],
            "init_liab_weight": args["init_liab_weight"],
            "liquidation_fee": args["liquidation_fee"],
            "maker_fee": args["maker_fee"],
            "taker_fee": args["taker_fee"],
            "min_funding": args["min_funding"],
            "max_funding": args["max_funding"],
            "impact_quantity": args["impact_quantity"],
            "group_insurance_fund": args["group_insurance_fund"],
            "trusted_market": args["trusted_market"],
            "fee_penalty": args["fee_penalty"],
            "settle_fee_flat": args["settle_fee_flat"],
            "settle_fee_amount_threshold": args["settle_fee_amount_threshold"],
            "settle_fee_fraction_low_health": args["settle_fee_fraction_low_health"],
            "settle_token_index": args["settle_token_index"],
        }
    )
    data = identifier + encoded_args
    return TransactionInstruction(keys, program_id, data)
